Client/Coordinators/ShareExtensionCoordinator.swift

```diff
@@ -0,0 +1,150 @@
+    func start(url: URL, sourceView: UIView, popoverArrowDirection: UIPopoverArrowDirection = .up) {
+        let shareExtension = ShareExtensionHelper(url: url, tab: tabManager.selectedTab)
+        let controller = shareExtension.createActivityViewController { [weak self] completed, activityType in
+            guard let self = self else { return }
+            self.handleShareExtensionCompletion(activityType: activityType)
+        }
+        if let popoverPresentationController = controller.popoverPresentationController {
+            popoverPresentationController.sourceView = sourceView
+            popoverPresentationController.sourceRect = sourceView.bounds
+            popoverPresentationController.permittedArrowDirections = popoverArrowDirection
+        }
+        TelemetryWrapper.recordEvent(category: .action, method: .tap, object: .sharePageWith)
+        router.present(controller)
+    }
+
+    private func handleShareExtensionCompletion(activityType: UIActivity.ActivityType?) {
+        switch activityType {
+        case CustomActivityAction.sendToDevice.actionType:
+            showSendToDevice()
+        case CustomActivityAction.copyLink.actionType:
+            SimpleToast().showAlertWithText(
+                .AppMenu.AppMenuCopyURLConfirmMessage,
+                bottomContainer: alertContainer,
+                theme: themeManager.currentTheme)
+            dequeueNotShownJSAlert()
+        default:
+            dequeueNotShownJSAlert()
+        }
+    }
+
+    private func showSendToDevice() {
+        guard let selectedTab = tabManager.selectedTab,
+              let url = selectedTab.canonicalURL?.displayURL
+        else { return }
+
+        let themeColors = themeManager.currentTheme.colors
+        let colors = SendToDeviceHelper.Colors(defaultBackground: themeColors.layer1,
+                                               textColor: themeColors.textPrimary,
+                                               iconColor: themeColors.iconPrimary)
+        let shareItem = ShareItem(url: url.absoluteString,
+                                  title: selectedTab.title)
+
+        let helper = SendToDeviceHelper(
+            shareItem: shareItem,
+            profile: profile,
+            colors: colors,
+            delegate: self)
+        let viewController = helper.initialViewController()
+
+        TelemetryWrapper.recordEvent(category: .action, method: .tap, object: .sendToDevice)
+        router.present(viewController)
+    }
+
+    private func dequeueNotShownJSAlert() {
+        guard let alertInfo = tabManager.selectedTab?.dequeueJavascriptAlertPrompt()
+        else {
+            parentCoordinator?.didFinish(from: self)
+            return
+        }
+        let alertController = alertInfo.alertController()
+        alertController.delegate = self
+        router.present(alertController)
+    }
+}
+
+extension ShareExtensionCoordinator: DevicePickerViewControllerDelegate, InstructionsViewDelegate {
```

Comment:
We prefer having the protocol conformance on the class definition rather than in an extension in the same file. This makes it implicit what the class is doing with a glance. (and then we add some `MARK: DevicePickerViewControllerDelegate`) where that specific code for that protocol starts.